File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
private:
    void solve(int idx, vector<int> &candidates, int target, vector<int> &curr, vector<vector<int>> &res) {
        // cout << "idx: " << idx << "\t target: " << target << endl;
        if (target == 0) {
            res.push_back(curr);
            return;
        }

        if (idx == candidates.size()) {
            return;
        }

        for (int i = idx; i < candidates.size(); ) {
            if (candidates[i] <= target) {
                curr.push_back(candidates[i]);
                solve(i + 1, candidates, target - candidates[i], curr, res);
                curr.pop_back();
                i++;

                while (i < candidates.size() && candidates[i] == candidates[i - 1]) {
                    i++;
                }
            } else {
                break;
            }
        }
    }

public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        int len = candidates.size();
        sort(candidates.begin(), candidates.end());

        vector<vector<int>> res;
        vector<int> curr;
        solve(0, candidates, target, curr, res);

        return res;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);

        Set<List<Integer>> ans = new HashSet<>();
        List<Integer> currElements = new ArrayList<>();
        solve(0, target, candidates, ans, currElements);

        List<List<Integer>> res = new ArrayList<>();
        for (List<Integer> aList : ans) {
            res.add(aList);
        }
        return res;
    }

    private void solve (int idx, int reqd, int[] candidates, Set<List<Integer>> ans, List<Integer> currElements) {
        if (reqd == 0) {
            ans.add(new ArrayList<>(currElements));
        }

        if (idx == candidates.length) {
            return;
        }

        if (candidates[idx] > reqd) {
            return;
        }

        // pick
        currElements.add(candidates[idx]);
        solve(idx + 1, reqd - candidates[idx], candidates, ans, currElements);

        // not pick - also remove last element added
        currElements.remove(currElements.size() - 1);
        solve(idx + 1, reqd, candidates, ans, currElements);
    }
}